### PR TITLE
fix: add revision check component to status metadata for email notification

### PR DESCRIPTION
**Pull request by mfrancisc**
Adding revision check conditions handling to the `ExtractStatusMetadata` function, so that it can be reported in the unready notification email.

### CODE REVIEW

File: controllers/toolchainstatus/toolchainstatus_controller_test.go [hunk 1]
```diff
@@ -1503,10 +1503,29 @@ func TestExtractStatusMetadata(t *testing.T) {
...
+		toolchainStatus.Status.HostOperator.RevisionCheck.Conditions, _ = condition.AddOrUpdateStatusConditions(
+			toolchainStatus.Status.HostOperator.Conditions, toolchainv1alpha1.Condition{
+				Type:    toolchainv1alpha1.ConditionReady,
+				Status:  corev1.ConditionFalse,
+				Reason:  toolchainv1alpha1.ToolchainStatusDeploymentNotUpToDateReason,
+				Message: "Deployed commit and GitHub commit are not matching",
+			})
+
+		meta := ExtractStatusMetadata(toolchainStatus)
+		require.Len(t, meta, 1)
```
Comment (MatousJobanek) [inline controllers/toolchainstatus/toolchainstatus_controller_test.go:1524]:
```suggestion
		// when
		meta := ExtractStatusMetadata(toolchainStatus)
		
		// then
		require.Len(t, meta, 1)
```

File: controllers/toolchainstatus/toolchainstatus_controller_test.go [hunk 2]
```diff
@@ -1564,6 +1583,26 @@ func TestExtractStatusMetadata(t *testing.T) {
...
+		toolchainStatus.Status.Members[0].MemberStatus.MemberOperator.RevisionCheck.Conditions, _ = condition.AddOrUpdateStatusConditions(
+			toolchainStatus.Status.Members[0].MemberStatus.MemberOperator.RevisionCheck.Conditions, toolchainv1alpha1.Condition{
+				Type:    toolchainv1alpha1.ConditionReady,
+				Status:  corev1.ConditionFalse,
+				Reason:  toolchainv1alpha1.ToolchainStatusDeploymentNotUpToDateReason,
+				Message: "Deployed commit and GitHub commit are not matching",
+			})
+
+		meta := ExtractStatusMetadata(toolchainStatus)
+		require.Len(t, meta, 1)
```
Comment (MatousJobanek) [inline controllers/toolchainstatus/toolchainstatus_controller_test.go:1599]:
```suggestion
		// when
		meta := ExtractStatusMetadata(toolchainStatus)
		
		// then
		require.Len(t, meta, 1)
```

File: controllers/toolchainstatus/toolchainstatus_controller_test.go [hunk 3]
```diff
@@ -1599,6 +1638,24 @@ func TestExtractStatusMetadata(t *testing.T) {
...
+		toolchainStatus := NewToolchainStatus(WithRegistrationService(WithRevisionCheckCondition(
+			toolchainv1alpha1.Condition{
+				Type:    toolchainv1alpha1.ConditionReady,
+				Status:  corev1.ConditionFalse,
+				Reason:  toolchainv1alpha1.ToolchainStatusDeploymentNotUpToDateReason,
+				Message: "Deployed commit and GitHub commit are not matching",
+			})))
+
+		meta := ExtractStatusMetadata(toolchainStatus)
+		require.Len(t, meta, 1)
```
Comment (MatousJobanek) [inline controllers/toolchainstatus/toolchainstatus_controller_test.go:1653]:
```suggestion
		// when
		meta := ExtractStatusMetadata(toolchainStatus)
		
		// then
		require.Len(t, meta, 1)
```

File: controllers/toolchainstatus/toolchainstatus_controller.go [hunk 4]
```diff
@@ -593,8 +593,18 @@ func ExtractStatusMetadata(instance *toolchainv1alpha1.ToolchainStatus) []*Compo
 		cond, found = condition.FindConditionByType(instance.Status.HostOperator.Conditions, toolchainv1alpha1.ConditionReady)
 		if found && cond.Status != corev1.ConditionTrue {
 			result = append(result, &ComponentNotReadyStatus{
-				ComponentType: "",
-				ComponentName: "Host Operator",
+				ComponentType: "Host Operator",
```
Comment (xcoulon) [inline controllers/toolchainstatus/toolchainstatus_controller.go:596]: shouldn't it be 
```
ComponentName: "Host Operator",
ComponentType: "Deployment",
```
instead?

File: controllers/toolchainstatus/toolchainstatus_controller.go [hunk 5]
```diff
@@ -593,8 +593,18 @@ func ExtractStatusMetadata(instance *toolchainv1alpha1.ToolchainStatus) []*Compo
...
+				ComponentName: "Deployment",
+				Reason:        cond.Reason,
+				Message:       cond.Message,
+			})
+		}
+
+		cond, found = condition.FindConditionByType(instance.Status.HostOperator.RevisionCheck.Conditions, toolchainv1alpha1.ConditionReady)
+		if found && cond.Status != corev1.ConditionTrue {
+			result = append(result, &ComponentNotReadyStatus{
+				ComponentType: "Host Operator",
```
Comment (xcoulon) [inline controllers/toolchainstatus/toolchainstatus_controller.go:606]: same here, `ComponentType` and `ComponentName` looks inverted

File: controllers/toolchainstatus/toolchainstatus_controller.go [hunk 6]
```diff
@@ -651,6 +673,16 @@ func ExtractStatusMetadata(instance *toolchainv1alpha1.ToolchainStatus) []*Compo
 				Message:       cond.Message,
 			})
 		}
+
+		cond, found = condition.FindConditionByType(instance.Status.RegistrationService.RevisionCheck.Conditions, toolchainv1alpha1.ConditionReady)
+		if found && cond.Status != corev1.ConditionTrue {
+			result = append(result, &ComponentNotReadyStatus{
+				ComponentType: "Registration service",
+				ComponentName: "Revision Check",
```
Comment (xcoulon) [inline controllers/toolchainstatus/toolchainstatus_controller.go:681]: same here, `ComponentType` and `ComponentName` looks inverted

File: controllers/toolchainstatus/toolchainstatus_controller.go [hunk 7]
```diff
@@ -628,6 +638,18 @@ func ExtractStatusMetadata(instance *toolchainv1alpha1.ToolchainStatus) []*Compo
 					})
 				}
 			}
+
+			if member.MemberStatus.MemberOperator != nil {
+				cond, found = condition.FindConditionByType(member.MemberStatus.MemberOperator.RevisionCheck.Conditions, toolchainv1alpha1.ConditionReady)
+				if found && cond.Status != corev1.ConditionTrue {
+					result = append(result, &ComponentNotReadyStatus{
+						ComponentType: "Member Revision Check",
```
Comment (xcoulon) [inline controllers/toolchainstatus/toolchainstatus_controller.go:646]: `ComponentType` value looks a bit weird. What about something like `Member Operator Revision` instead?

**Review by MatousJobanek — APPROVED** [review]

File: controllers/toolchainstatus/toolchainstatus_controller_test.go [hunk 8]
```diff
@@ -1642,7 +1699,7 @@ func TestGenerateUnreadyNotificationContent(t *testing.T) {
 		require.Len(t, meta, 4)
 		content, err := GenerateUnreadyNotificationContent(ClusterURLs(toolchainStatus), meta)
 		require.NoError(t, err)
-		require.Len(t, content, 1512)
+		require.Len(t, content, 1522)
```
Comment (MatousJobanek) [inline controllers/toolchainstatus/toolchainstatus_controller_test.go:1702]: I guess that counting the length of the string is not so important. We could have this instead:
```suggestion
		require.NotEmpty(t, content)
		assert.Contains(t, content, "<h4>Host Operator Deployment not ready</h4>")
		assert.Contains(t, content, "<h4>Member Routes member-sandbox.ccc.openshiftapps.com not ready</h4>")
		assert.Contains(t, content, "<h4>Registration service deployment not ready</h4>")
```

File: controllers/toolchainstatus/toolchainstatus_controller.go (hunk 4 above)
Comment (mfrancisc) [reply controllers/toolchainstatus/toolchainstatus_controller.go:596]: TBH I just followed the already exiting convention where component type was either: "Host Operator", "Member" or "Registration Service". I can Invert those , but if we do this we need to invert also the other existing occurrences not only the ones I've added/changed right ?

Comment (mfrancisc) [reply controllers/toolchainstatus/toolchainstatus_controller.go]: inverted , please see: https://github.com/codeready-toolchain/host-operator/pull/836/commits/54d573742cad4df097d7652a5ae06a356182acf0

File: controllers/toolchainstatus/toolchainstatus_controller.go (hunk 5 above)
Comment (mfrancisc) [reply controllers/toolchainstatus/toolchainstatus_controller.go]: inverted , please see: https://github.com/codeready-toolchain/host-operator/pull/836/commits/54d573742cad4df097d7652a5ae06a356182acf0

File: controllers/toolchainstatus/toolchainstatus_controller.go (hunk 6 above)
Comment (mfrancisc) [reply controllers/toolchainstatus/toolchainstatus_controller.go]: inverted , please see: https://github.com/codeready-toolchain/host-operator/pull/836/commits/54d573742cad4df097d7652a5ae06a356182acf0

File: controllers/toolchainstatus/toolchainstatus_controller.go (hunk 7 above)
Comment (mfrancisc) [reply controllers/toolchainstatus/toolchainstatus_controller.go]: applied, please see: https://github.com/codeready-toolchain/host-operator/pull/836/commits/54d573742cad4df097d7652a5ae06a356182acf0

File: controllers/toolchainstatus/toolchainstatus_controller_test.go (hunk 8 above)
Comment (mfrancisc) [reply controllers/toolchainstatus/toolchainstatus_controller_test.go]: Agree , it makes the test a bit more flexible , applied please see: https://github.com/codeready-toolchain/host-operator/pull/836/commits/54d573742cad4df097d7652a5ae06a356182acf0

**Review by xcoulon — APPROVED** [review]
thanks for addressing my comments, @mfrancisc!

**Review by MatousJobanek — APPROVED** [review]

**Review by rajivnathan — APPROVED** [review]
Looks good. Shouldn't the `<h4>{{.ComponentType}} {{.ComponentName}} not ready</h4>` in `statusNotificationTemplate` also be inverted?

**Comment by mfrancisc** [comment]
@rajivnathan I did that in https://github.com/codeready-toolchain/host-operator/pull/836/commits/2e0b66db9df53530166212f50e2e7b347d70b206, but it still looks a bit "weird" IMO for the toolchain status string that has an extra space now:
_"ToolchainStatus  not ready"_

and the member components:
_"member-sandbox.ccc.openshiftapps.com Member Routes not ready"_

but I wasn't able to find a better formatting for now.

**Pull request merged**